server/src/main/java/de/uftos/repositories/ucdl/parser/DefinitionParser.java

```diff
@@ -0,0 +1,691 @@
+          case "RETURN":
+            ValueDto<Boolean> dto =
+                (ValueDto<Boolean>) buildAst(bodyElements.jjtGetChild(0), parameters);
+            returnValue = dto.value();
+            body.add(dto);
+            break;
+          case "CONTROLSEQUENCE":
+            body.add(buildAst(bodyElements.jjtGetChild(0), parameters)); //first control sequence
+
+            Node controlSequenceList = bodyElements.jjtGetChild(1); //rest of the control sequences
+            while (controlSequenceList.jjtGetNumChildren() == 2) {
+              body.add(buildAst(controlSequenceList.jjtGetChild(0), parameters));
+              controlSequenceList = controlSequenceList.jjtGetChild(1);
+            }
+            //semantic check for legal return values
+            returnValue = getReturnValue(body);
+            break;
+          default:
+            throw new IllegalStateException();
+        }
+        return new ControlSequenceDto(UcdlToken.IF, bool, body, returnValue);
+      }
+      case "FORALL" -> {
+        List<AbstractSyntaxTreeDto> variableDefinition = new ArrayList<>();
+        ValueDto<String> varName = (ValueDto<String>) buildAst(root.jjtGetChild(0), parameters);
+        variableDefinition.add(varName); //variable name
+        if (parameters.containsKey(varName.value())) {
+          throw new ParseException("Variable name \"" + varName.value()
+              + "\" does already exist inside this namespace!");
+        }
+        SetDto set = (SetDto) buildAst(root.jjtGetChild(1), parameters);
+        variableDefinition.add(set); //set
+        OperatorDto variable = new OperatorDto(UcdlToken.OF, variableDefinition);
+
+        parameters.put(varName.value(), set.type()); //adding the new variable
+        AbstractSyntaxTreeDto body = buildAst(root.jjtGetChild(1), parameters);
+        parameters.remove(varName.value()); //removing the new variable
+
+        return new QuantifierDto(UcdlToken.FOR_ALL, variable, body);
+      }
+      case "EXISTS" -> {
+        List<AbstractSyntaxTreeDto> variableDefinition = new ArrayList<>();
+        ValueDto<String> varName = (ValueDto<String>) buildAst(root.jjtGetChild(0), parameters);
+        variableDefinition.add(varName); //variable name
+        if (parameters.containsKey(varName.value())) {
+          throw new ParseException("Variable name \"" + varName.value()
+              + "\" does already exist inside this namespace!");
+        }
+        SetDto set = (SetDto) buildAst(root.jjtGetChild(1), parameters);
+        variableDefinition.add(set); //set
+        OperatorDto variable = new OperatorDto(UcdlToken.OF, variableDefinition);
+
+        parameters.put(varName.value(), set.type()); //adding the new variable
+        AbstractSyntaxTreeDto body = buildAst(root.jjtGetChild(1), parameters);
+        parameters.remove(varName.value()); //removing the new variable
+
+        return new QuantifierDto(UcdlToken.EXISTS, variable, body);
+      }
```

Comment:
Those two are identical except for the very last line. That's generally a pattern here; that code blocks are copied multiple times. It definitely makes sense to refactor that at some point.